### PR TITLE
GH-1477 Wildcard/Regex Whitelisting

### PR DESCRIPTION
**Pull request by benstrumeyer**
* [*] Have you followed the guidelines in [CONTRIBUTING.md](../CONTRIBUTING.md)?
* [*] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?
* [*] Have you added an explanation of what your changes do?
* [*] Does your submission pass tests?
* [*] Did you lint your code prior to submission?

- Add wildcard and regex input for whitelist/blacklist
- Does not include paygating
- Add UI and background unit tests
- Add regex-safe dependency (https://www.npmjs.com/package/safe-regex)
- Add jest-when dependency (https://www.npmjs.com/package/jest-when)

Ticket: https://cliqztix.atlassian.net/browse/GH-1477

### CODE REVIEW

File: src/classes/Policy.js [hunk 1]
```diff
@@ -174,6 +175,34 @@ class Policy {
...
+
+	/**
+	 * Check given url against pattern which might be a wildcard, or a regex
+	 * @param  {string} url		site url
+	 * @param  {string} pattern	regex pattern
+	 * @return {boolean}
+	 */
+	matchesWildcardOrRegex(url, pattern) {
+		// Pattern might be a wildcard
+		if (pattern.includes('*')) {
```
Comment (christophertino) [inline src/classes/Policy.js]: On fresh install, this is throwing a background error `Cannot read property 'includes' of null`

Comment (christophertino) [reply src/classes/Policy.js]: Looks to be coming from BrowserButton ln 154 `this._setIcon(!globals.SESSION.paused_blocking && !this.policy.checkSiteWhitelist(tab.url), tabId, trackerCount, alert);`

File: app/panel/components/Settings/TrustAndRestrict.jsx [hunk 2]
```diff
@@ -139,6 +140,44 @@ class TrustAndRestrict extends React.Component {
 		this.props.actions.updateSitePolicy({ type: listType, pageHost });
 	}
```
Comment (christophertino) [inline app/panel/components/Settings/TrustAndRestrict.jsx]: Once the site is successfully added to the list, clear the input field

File: src/classes/Policy.js (hunk 1 above)
Comment (benstrumeyer) [reply src/classes/Policy.js]: Nice catch, I've added a check to ensure pattern is truthy

File: app/panel/components/Settings/TrustAndRestrict.jsx (hunk 2 above)
Comment (benstrumeyer) [reply app/panel/components/Settings/TrustAndRestrict.jsx]: Input field cleared!

File: app/panel/components/Settings/TrustAndRestrict.jsx [hunk 3]
```diff
@@ -137,6 +137,35 @@ class TrustAndRestrict extends React.Component {
...
 		}
 		this.props.actions.updateSitePolicy({ type: listType, pageHost });
+		if (listType === 'whitelist') {
+			this.setState({ trustedValue: '' });
+		} else {
+			this.setState({ restrictedValue: '' });
+		}
+	}
+
+	isValidUrlorWildcard(pageHost) {
```
Comment (christophertino) [inline app/panel/components/Settings/TrustAndRestrict.jsx]: This fails when trying to add `localhost:3000` to the list.

Comment (benstrumeyer) [reply app/panel/components/Settings/TrustAndRestrict.jsx]: Good catch! I fixed the validation regex and added a unit test

**Pull request merged**